Unref task after return

danigm opened the pull request:
https://phabricator.endlessm.com/T29887

andrunko commented:
@pwithnall could you help trying to understand why this seems to be needed? See also https://github.com/endlessm/eos-hack-extension/pull/16#discussion_r410179888.

pwithnall committed:
animations-dbus: Fix double-unref of GDBusConnection

Spotted while reading the code. This might never have manifested in a
bug because it would have been the global session `GDBusConnection`,
which normally has several other refs anyway.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T29887

pwithnall committed:
animations-dbus: More carefully track GTask ownership transfer

Add `g_steal_pointer()` calls in a few places to clarify the ownership
transfer of a reference to the `GTask`. Add a few `g_autoptr()` calls to
ensure that the reference is dropped in all the appropriate places too —
this fixes two refcount leaks on two paths.

Spotted and diagnosed by Daniel Garcia Moreno.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T29887

pwithnall commented on animations-dbus/animations-dbus-server-object.c:
```diff
@@ -454,7 +456,7 @@ attempt_to_own_session_bus_name (GDBusConnection *connection,
                                        on_lost_session_bus_name,
                                        task,
-                                       g_object_unref);
+                                       NULL);
```
I think this will introduce some new bugs: `g_dbus_own_name_on_connection()` is allowed to call the name-acquired and name-lost callbacks multiple times, so dropping the reference to the `GTask` after one of them completes means there will be double- and triple-unrefs in future. I think the original code here is correct, and [the bug](https://github.com/endlessm/eos-hack-extension/pull/16#discussion_r410179888) is somewhere else.

pwithnall replied:
(Outdated.)

pwithnall commented on animations-dbus/animations-dbus-server-object.c:
```diff
@@ -471,6 +473,7 @@ on_got_session_bus_connection (GObject      *object G_GNUC_UNUSED,
   if (connection == NULL)
     {
       g_task_return_error (task, g_steal_pointer (&local_error));
+      g_object_unref (task);
```
This (and the one below) look like valid fixes. I have a slightly more comprehensive patch which cleans up the code a bit (I wrote it while trying to understand the reference count transfers), which you might want to use/incorporate/ignore depending on whether you think it’s correct. I haven’t tested it (and since I’m just doing a drive-by review here, I have no plans to take it any further):

https://github.com/endlessm/libanimations-dbus/tree/T29817-refcounting

pwithnall replied:
(Outdated.)

pwithnall reviewed (commented):
Are you sure the `g_object_unref()` critical warning you saw [in this comment](https://github.com/endlessm/eos-hack-extension/pull/16#discussion_r410179888) was for the `GTask`? I can’t see any way for that to happen with the existing code.

In any case, in looking at this code you’ve detected a few bugs, so hopefully we can make progress by fixing those.

pwithnall reviewed (dismissed):
(needs changes as per above comments)

pwithnall committed:
server-object: Fix a typo in a developer comment

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T29887

pwithnall committed:
animation-manager: Add an unexport() method

This mirrors the `export()` method, and will be used in an upcoming
commit.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T29887

pwithnall committed:
server-object: Factor out an unregister_client() helper method

This introduces no functional changes, but will make some upcoming
changes simpler.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T29887

pwithnall committed:
server-object: Unexport manager objects when clients disconnect

When a client disconnects, the manager object exported for it should be
removed.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T29887

pwithnall committed:
server-object: Add a stop() method

This will be used to allow the D-Bus service to be stopped (all objects
unexported and the D-Bus well known name unowned) without having to
reach finalisation of the server object.

Tying finalisation of the object to unowning the name makes it difficult
to guarantee the name is unowned when the calling code’s final reference
to the server is dropped, as another bit of code (or the garbage
collector) may still hold a reference.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T29887

pwithnall commented:
OK, I’ve force-pushed some updates over this branch, and the PR needs re-reviewing by someone other than me. The new branch incorporates Daniel’s original fixes.

andrunko commented on animations-dbus/animations-dbus-server-object.c:
```diff
@@ -735,3 +744,68 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
+      AnimationsDbusServerSurface *surface = g_ptr_array_index (priv->animatable_surfaces, i);
+
+      if (!animations_dbus_server_unregister_surface (self, surface, NULL))
+        error_seen = TRUE;
```
Here you are iterating while removing from the same array, isn't that wrong given `array->len` is decreasing while the index (`i`) is increasing at the same time? I think a while loop would probably be clearer:
```
while (priv->animatable_surfaces->len > 0)
  {
     AnimationsDbusServerSurface *surface = g_ptr_array_index (priv->animatable_surfaces, 0);

      if (!animations_dbus_server_unregister_surface (self, surface, NULL))
        error_seen = TRUE;
  }
```

pwithnall replied:
Good catch, fixed.

andrunko commented on animations-dbus/animations-dbus-server-object.c:
```diff
@@ -735,3 +744,68 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
+{
+  AnimationsDbusServerPrivate *priv = animations_dbus_server_get_instance_private (self);
+  guint i;
+  GHashTableIter iter;
```
I guess we could declare `iter` together with `key` in the conditional below.

pwithnall replied:
Fixed.

andrunko commented on animations-dbus/animations-dbus-server-object.c:
```diff
@@ -735,3 +744,68 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
+gboolean
+animations_dbus_server_stop (AnimationsDbusServer  *self,
+                             GCancellable          *cancellable,
+                             GError               **error)
```
Can we restart the server after calling `stop()`? If not, probably worth a note in the docs. Alternatively we could make this private and only invoked from dispose - for the hack extension we could force a call to `animationDbusServer.run_dispose()` when stopping it.

pwithnall replied:
No, it can’t be restarted. I’ve added that to the documentation. Forcing a call to `run_dispose()` is almost always a hack (because it interferes with the strong refs still potentially held by other code) so I’m not going to do that. I think it’s fine for the object to essentially be useless (but harmless) after `stop()` is called.

andrunko reviewed (changes requested):
Some minor comments, otherwise lgtm.

andrunko commented on animations-dbus/animations-dbus-server-object.c:
```diff
@@ -735,3 +744,70 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
+        {
+          const gchar *name = key;
+
+          unregister_client (self, name);
```
Sorry, missed this in the prev review - wouldn't that also give us issues with removing while iterating? The doc for `g_hash_table_iter_init` says:
```
Initializes a key/value pair iterator and associates it with hash_table . Modifying the hash table after calling this function invalidates the returned iterator.
```

pwithnall replied:
Fixed

andrunko reviewed (commented):
Left another small comment, otherwise lgtm.

andrunko commented on animations-dbus/animations-dbus-server-object.c:
```diff
@@ -735,3 +744,69 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
+  if (priv->animation_managers != NULL)
+    {
+      /* Operate on a copy of the keys of @animation_managers to avoid iterating while modifying the hash table */
+      GList *names = g_hash_table_get_keys (priv->animation_managers);
```
I dont think that is the case, but wonder if the contents of the keys can be modified once the hash is modified. From docs:
```
The content of the list is owned by the hash table and should not be modified or freed. Use g_list_free() when done using the list.
```

Alternatively we could use g_hash_table_get_keys_as_array which seems to always give us a copy?

pwithnall replied:
Uff, the return value of `g_hash_table_get_keys()` is of course `(transfer container)` and not `(transfer full)`. Fixed that.

The individual name strings should be OK to use even while the table is modified, since only the one which we’re calling `unsubscribe_client()` on should end up being freed, and that’s after this code is finished with it.

pwithnall replied:
`g_hash_table_get_keys_as_array()` can’t give us a copy as `GHashTable` doesn’t know how to deep-copy the key or value pointers it stores.

andrunko replied:
Great, I had missed the `g_list_free_full()` issue, good catch.

andrunko reviewed (approved):
LGTM.
